### PR TITLE
Overtime vaults

### PR DESCRIPTION
**Pull request by danijelthales**
Reworked vaults approach for Overtime

### CODE REVIEW

File: contracts/OvertimeVaults/BaseSportVault.sol [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    uint public utilizationRate;
+
+    mapping(uint => uint) public capPerRound;
+
+    uint public minDepositAmount;
+
+    uint public maxAllowedUsers;
+    uint public usersCurrentlyInVault;
+
+    //TODO: add staking thales so gamified staking bonuses are updated on every round closing
```
Comment (grujawork) [inline contracts/OvertimeVaults/BaseSportVault.sol:64]: this is not implemented?

File: contracts/OvertimeVaults/BaseSportVault.sol [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    /// @notice Set length of rounds
+    /// @param _roundLength Length of a round in miliseconds
+    function setRoundLength(uint _roundLength) external onlyOwner {
+        roundLength = _roundLength;
+        emit RoundLengthChanged(_roundLength);
+    }
+
+    /// @notice Set ThalesAMM contract
+    /// @param _sportAMM ThalesAMM address
+    function setSportAmm(ISportsAMM _sportAMM) external onlyOwner {
```
Comment (grujawork) [inline contracts/OvertimeVaults/BaseSportVault.sol:208]: address zero?

File: contracts/OvertimeVaults/BaseSportVault.sol [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            capPerRound[nextRound] -= balancesPerRound[round][msg.sender];
+        }
+
+        withdrawalRequested[msg.sender] = true;
+        emit WithdrawalRequested(msg.sender);
+    }
+
+    /// @notice Set length of rounds
+    /// @param _roundLength Length of a round in miliseconds
+    function setRoundLength(uint _roundLength) external onlyOwner {
```
Comment (grujawork) [inline contracts/OvertimeVaults/BaseSportVault.sol:201]: maybe to add like owner method tests to increase coverage to max :) 

See coverage: https://app.codecov.io/gh/thales-markets/contracts/pull/219

File: contracts/OvertimeVaults/BaseSportVault.sol [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        emit MaxAllowedUsersChanged(_maxAllowedUsers);
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    function _exerciseMarketsReadyToExercised() internal {
+        ISportPositionalMarket market;
+        for (uint i = 0; i < tradingMarketsPerRound[round].length; i++) {
+            market = ISportPositionalMarket(tradingMarketsPerRound[round][i]);
+            if (!market.paused()) {
```
Comment (grujawork) [inline contracts/OvertimeVaults/BaseSportVault.sol:248]: maybe to combine in && with below

Comment (grujawork) [reply contracts/OvertimeVaults/BaseSportVault.sol:248]: market.resolved() && !market.paused()

File: contracts/OvertimeVaults/BaseSportVault.sol [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        for (uint i = 0; i < tradingMarketsPerRound[round].length; i++) {
+            ISportPositionalMarket market = ISportPositionalMarket(tradingMarketsPerRound[round][i]);
+            if ((!market.resolved()) || market.paused()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    //    function hasMarketsReadyToExercise() external view returns (bool) {
```
Comment (grujawork) [inline contracts/OvertimeVaults/BaseSportVault.sol:280]: remove

File: contracts/OvertimeVaults/SportVault.sol [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+        if (!isTradingMarketInARound[round][market]) {
+            tradingMarketsPerRound[round].push(market);
+            isTradingMarketInARound[round][market] = true;
+        }
+    }
+
+    /// @notice Set allocation limits for assets to be spent in one round
+    /// @param _allocationLimitsPerMarketPerRound allocation per market in percent
+    function setAllocationLimits(uint _allocationLimitsPerMarketPerRound) external onlyOwner {
```
Comment (grujawork) [inline contracts/OvertimeVaults/SportVault.sol:95]: also to cover by test so increase % of tests all methods of owners

**Review by grujawork — COMMENTED** [review]
All good, generally i look at PR on code coverage and see where it is read so i can cover

PR  code coverage: https://app.codecov.io/gh/thales-markets/contracts/pull/219

File: contracts/OvertimeVaults/BaseSportVault.sol (hunk 4 above)
Comment (danijelthales) [reply contracts/OvertimeVaults/BaseSportVault.sol:245]: done

File: contracts/OvertimeVaults/BaseSportVault.sol (hunk 5 above)
Comment (danijelthales) [reply contracts/OvertimeVaults/BaseSportVault.sol:277]: done

File: contracts/OvertimeVaults/BaseSportVault.sol (hunk 3 above)
Comment (danijelthales) [reply contracts/OvertimeVaults/BaseSportVault.sol:198]: done

File: contracts/OvertimeVaults/BaseSportVault.sol (hunk 1 above)
Comment (danijelthales) [reply contracts/OvertimeVaults/BaseSportVault.sol:64]: no, it needs more thought, but I dont want to forget about it :)

**Review by grujawork — APPROVED** [review]

**Pull request merged**